[DOCS] WIP Docs/custom expectation test cases (#3969)
great-expectations/great_expectations: docs/guides/expectations/features_custom_expectations/how_to_add_example_cases_for_an_expectation.md

---
title: How to create example cases for a Custom Expectation
---
import Prerequisites from '../creating_custom_expectations/components/prerequisites.jsx'

This guide will help you add example cases to document and test the behavior of your [Expectation](../../../reference/expectations/expectations.md). 

<Prerequisites>

 - Created a [Custom Expectation](../creating_custom_expectations/overview.md)

</Prerequisites>

> Example cases in Great Expectations serve a dual purpose:
> * They help the users of the Expectation understand its logic by providing examples of input data that the Expectation will evaluate;
> * And provide test cases that the Great Expectations testing framework can execute automatically.

Reviewer: ```suggestion
Example cases in Great Expectations serve a dual purpose:
* First, they help the users of the Expectation understand its logic by providing examples of input data that the Expectation will evaluate.
* Second, they provide test cases that the Great Expectations testing framework can execute automatically.
```
Since you state a dual purpose, associating each item with a number can help people keep things mentally sorted.  Likewise, the connection between the two items is implied by their grouping in the list -- you can just treat each as a full sentence, instead of using semicolon list notation, which is more formal than our style guide calls for.

If you decide to contribute your Expectation, its entry in the [Expectations Gallery](https://greatexpectations.io/expectations/) will render these examples.

We will explain the structure of these tests using the Custom Expectation implemented in [How to create Custom Column Aggregate Expectations](../creating_custom_expectations/how_to_create_custom_column_aggregate_expectations.md).

> ### Steps
> 
> #### 1. Decide which tests you want to implement

Reviewer: ```suggestion
## Steps

### 1. Decide which tests you want to implement
```
As a rule of thumb we want our headings to nest from H1 > H2 > H3. Since the page's title is treated as our H1, the next major subsection should be at the H2 level, and the subsections under that should be at H3.

<code>{`{
  "success": True,
  "expectation_config": {
      "expectation_type": "expect_column_value_z_scores_to_be_less_than",
      "kwargs": {
          "column": "a",
          "mostly": 0.9,
          "threshold": 4,
          "double_sided": True,
      },
      "meta": {},
  },
  "result": {
      "element_count": 6,
      "unexpected_count": 0,
      "unexpected_percent": 0.0,
      "partial_unexpected_list": [],
      "missing_count": 0,
      "missing_percent": 0.0,
      "unexpected_percent_total": 0.0,
      "unexpected_percent_nonmissing": 0.0,
  },
  "exception_info": {
      "raised_exception": False,
      "exception_traceback": None,
      "exception_message": None,
  }
}`}
</code>
</details>

At a minimum, we want to create tests that show what our Custom Expectation will and will *not* do. 
These basic positive and negative example cases are the minimum amount of test coverage required for a Custom Expectation to be accepted into the Great Expectations codebase at an [Experimental level](../../../contributing/contributing_maturity.md#contributing-expectations).

> To begin with, let's implement those two basic tests; one positive example case, and one negative example case. 

Reviewer: ```suggestion
To begin with, let's implement those two basic tests: one positive example case, and one negative example case. 
```
A ; is usually only appropriate if it can be replaced with the word "and" and have the sentence still make sense.  On the other hand, a colon can be used to denote that the following is a list of some sort.

> #### 2. Defining our data

Reviewer: ```suggestion
### 2. Defining our data
```

> Search for `examples = []` in the template file you are modifying for your new custom Expectation. 

Reviewer: ```suggestion
Search for `examples = []` in the template file you are modifying for your new Custom Expectation. 
```
Technically "Custom Expectation" is a technical term associated with "Expectation" -- it denotes a specific type of Expectation that will eventually have its own documentation.  Therefore the entire term "Custom Expectation" should be title cased.

        <li> <code>exact_match_out</code>: if you set <code>exact_match_out=False</code>, then you don’t need to include all the elements of the result object - only the ones that are important to test, such as <inlineCode>{`{"success": True}`}</inlineCode>. </li>
    </ul>
</ul>
</details>

In our example, `data` will have two columns, "x" and "y", each with five rows. 
If you define multiple columns, make sure that they have the same number of rows. 
When possible, include test data and tests that includes null values (`None` in the Python test definition).

```python file=../../../../tests/integration/docusaurus/expectations/creating_custom_expectations/expect_column_max_to_be_between_custom.py#L46
```

:::note
When you define data in your examples, we will mostly guess the type of the columns. 
Sometimes you need to specify the precise type of the columns for each backend. Then you use the `schema` attribute (on the same level as `data` and `tests` in the dictionary):

```console
"schemas": {
  "spark": {
    "x": "IntegerType",
  },
  "sqlite": {
    "x": "INTEGER",
  },
```
:::

> #### 3. Defining our tests

Reviewer: ```suggestion
### 3. Defining our tests
```

In our example, `tests` will be a list containing dictionaries defining each test. 

You will need to:

> 1. Title your tests (`title`);
> 2. Define the input for your tests (`in`);
> 3. Decide how precisely you want to test the output of your tests (`exact_match_out`);
> 4. Define the expected output for your tests (`out`).

Reviewer: ```suggestion
1. Title your tests (`title`).
2. Define the input for your tests (`in`).
3. Decide how precisely you want to test the output of your tests (`exact_match_out`).
4. Define the expected output for your tests (`out`).
```
Semicolon formatting of ordered lists is more formal than our style guide requires. In general, punctuation is not required for list items that are fragments, while sentences should start with a capital and end with a period. (And ideally a list will contain all sentences or all fragments as its items.)

If you are interested in contributing your Custom Expectation back to Great Expectations, you will also need to decide if you want these tests publically displayed to demonstrate the functionality of your Custom Expectation (`include_in_gallery`).

```python file=../../../../tests/integration/docusaurus/expectations/creating_custom_expectations/expect_column_max_to_be_between_custom.py#L44-L76
```

> #### 4. Verifying our tests

Reviewer: ```suggestion
### 4. Verifying our tests
```

If you now run your file, `print_diagnostic_checklist` will attempt to execute these example cases.

If the tests are correctly defined, and the rest of the logic in your Custom Expectation is already complete,
you will see the following in your Diagnostic Checklist:

```console
✔ Has at least one positive and negative example case, and all test cases pass
```

<div style={{"text-align":"center"}}>
<p style={{"color":"#8784FF","font-size":"1.4em"}}><b>
Congratulations!<br/>&#127881; You've successfully created example cases & tests for a Custom Expectation! &#127881;
</b></p>
</div>

> #### 5. Contribution (Optional)

Reviewer: ```suggestion
### 5. Contribution (Optional)
```

> This guide will leave you with test coverage sufficient for [contribution](../contributing/how_to_contribute_a_new_expectation_to_great_expectations.md) back to Great Expecations at an Experimental level;
> if you're interested in having your contribution accepted at a Beta level, these tests will need to pass for all supported backends (Pandas, Spark, & SQLAlchemy).

Reviewer: ```suggestion
This guide will leave you with test coverage sufficient for [contribution](../contributing/how_to_contribute_a_new_expectation_to_great_expectations.md) back to Great Expecations at an Experimental level.  

If you're interested in having your contribution accepted at a Beta level, these tests will need to pass for all supported backends (Pandas, Spark, & SQLAlchemy).
```
While a semicolon technically would work to separate these two sentences, it is more consistent to treat them as individual statements since the third scenario -- full acceptance into the Great Expectations codebase at Production level -- is an individual statement.  It is also generally easier for users to parse shorter sentences, so splitting this into two discrete statements increases readability.

For full acceptance into the Great Expectations codebase at a Production level, we require a more robust test suite. 
If you believe your Custom Expectation is otherwise ready for contrbution at a Production level, please submit a [Pull Request](https://github.com/great-expectations/great_expectations/pull-requests), and we will work with you to ensure adequate testing.

> :::note
> For more information on our code standards and contribution, see our guide on [Levels of Maturity](../../../contributing/contributing_maturity.md#contributing-expectations) for Expectations
> :::

Reviewer: ```suggestion
:::note
For more information on our code standards and contribution, see our guide on [Levels of Maturity](../../../contributing/contributing_maturity.md#contributing-expectations) for Expectations.

To view the full script used in this page, see it on GitHub:
- [expect_column_max_to_be_between_custom.py](https://github.com/great-expectations/great_expectations/blob/hackathon-docs/tests/integration/docusaurus/expectations/creating_custom_expectations/expect_column_max_to_be_between_custom.py)
:::
```
Good job on using file and reference numbers for the code examples!  Our documentation standards are such that at the end of a document that does this (as all documents with code examples ideally should) we also want to link to the full code so the user can reference it in its entirety if they need to.  You may need to adjust the link that is used in this suggestion: it is pointing to the file in the hackathon-docs branch, since that file doesn't exist at that location in other branches currently.
